### PR TITLE
#65 support inherit configuration

### LINKED ISSUE #65: Support for the @InheritConfiguration
Let's say we have an `A` class. For it there is `ADto` class. 
We create a mapping `ADto A_to_ADto(A source)` and specify all the mappings with `@Mapping` annotation.
Now we want `ASpecialDto` class that extends `ADto` and adds some new props that weren't mapped before.
We use `@InheritConfiguration` on a new mapping `ASpecialDto A_to_ASpecialDto(A source)` that uses previous mapping for our needs.
The plugin can't detect that we are inheriting the config and proposes to define mapping once again.

This issue may correlate with #35 
image

### PR DESCRIPTION
**Pull request by thunderhook**
I could narrow it down to the following use cases:
* `@InheritConfiguration` from a method in...
    * the same mapper (local reference)
    * super mapper
    * by name 
    * a referenced `@MapperConfig`
    * the same mapper (local reference) but with another type that can be inherited from (e.g. inheritance like `Person <- NaturalPerson`)

For all these I have provided a test case.

I am using an approach with two steps:
* search for all possible methods in the scope where `@InheritConfiguration` might be possible
* find a single method that can be inherited from - as you suggested: when there is more than one candidate, nothing will be returned

Fixes #65

### CODE REVIEW

File: src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        return Stream.concat( Stream.concat( localAndParentMethods, referencedMethods ), mapperConfigMethods );
+    }
+
+    private static List<PsiMethod> findLocalAndParentMethods(PsiClass containingClass) {
+
+        List<PsiMethod> result = new ArrayList<>( List.of( containingClass.getMethods() ) );
+
+        for ( PsiClass anInterface : containingClass.getInterfaces() ) {
+            result.addAll( findLocalAndParentMethods( anInterface ) );
+        }
```
Comment (filiphr) [inline src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java:73]: Can't we use `PsiClass#getAllMethods`?

File: src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+        return candidate.getName().equals( inheritConfigurationName );
+    }
+
+    /**
+     * simplified version of <code>org.mapstruct.ap.internal.model.source.SourceMethod#canInheritFrom</code>
+     */
+    public static boolean canInheritFrom(PsiMethod mappingMethod, PsiMethod candidate) {
+
+        PsiType targetType = findTargetTypeOfMappingMethod( mappingMethod );
```
Comment (filiphr) [inline src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java:136]: Slight optimization. Instead of getting the target type every time, why not pass it to the `canInheritMethod` instead?

File: src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            .filter( candidate -> isNotTheSameMethod( mappingMethod, candidate ) )
+            .filter( candidate -> matchesNameWhenNameIsDefined( inheritConfigurationName, candidate ) )
+            .filter( candidate -> canInheritFrom( mappingMethod, candidate ) )
+            .collect( Collectors.toList() );
+
+        if ( matchingCandidates.size() != 1 ) {
+            return Optional.empty();
+        }
+
+        return matchingCandidates.stream().findFirst();
```
Comment (filiphr) [inline src/main/java/org/mapstruct/intellij/util/InheritConfigurationUtils.java:114]:
```suggestion
        if ( matchingCandidates.size() == 1 ) {
            return Optional.of( matchingCandidates.get( 0 ) );
        }

        return Optional.empty();
```

**Review by filiphr — COMMENTED** [review]
Thanks @thunderhook. This looks good. I've left some small comments.

Apart from those, can we perhaps add a test for inheriting configuration in the reverse. i.e. `@Mapping` defined on update methods and `@InheritConfiguration` on the normal mapping? We only need one, we don't need all the scenarios for this.

**Comment by thunderhook** [comment]
Thank you for the review!

Actually the inheriting in reverse did not work, thank you for the suggestion to test it.
I also stumbled accross the edge case using a `@Context` variable in a method to inherit from and used `MapstructUtil::isValidSourceParameter` for checking the equality of the parameter list.

Let me know if something else needs to be changed.

**Comment by filiphr** [comment]
@thunderhook I thought I wrote it in my previous review, but it seems that I didn't. Can we also include `BeanMapping#unmappedTargetPolicy` here as well? It should have precedence over the `@Mapper`

**Comment by thunderhook** [comment]
@filiphr Just to clarify the use case:

So that an incomplete mapping like `carDtoToCar` should also show no warning when it is used to inherit from?

```java
@Mapper(unmappedTargetPolicy = ReportingPolicy.ERROR) // or WARN
public interface InheritConfigurationInSameClassBeanMappingMapper {

    @Mapping(target = "manufacturingDate", source = "manufacturingYear")
    @Mapping(target = "numberOfSeats", source = "seatCount")
    // @Mapping(target = "free", source = "available")
    // @Mapping(target = "driver", source = "myDriver")
    @BeanMapping(unmappedTargetPolicy = ReportingPolicy.IGNORE)
    Car carDtoToCar(CarDto car);

    @Mapping(target = "free", constant = "false")
    @InheritConfiguration
    void carDtoIntoCar(CarDto carDto, @MappingTarget Car car); // currently this warns about the missing target "driver"
    
}
```

This currently (with the implemented branch) shows a warning about the missing target "driver".

The current approach searches for all needed target properties and removes all (directly or indirectly defined) mapped properties to show the remaining (missing) target properties in the warning.
I can't remove "driver" here because it isn't defined anywhere.
Is it safe to just disable the unmappedTargetProperty warning altogether (irrespective of which target properties are missing) when the matching candidate method contains `@BeanMapping(unmappedTargetPolicy = ReportingPolicy.IGNORE)`?

Or did I misunderstand you? 

Did you mean to provide the option to disable the warning or set it as an error as in pull request #131? Which kind of collides currently (if you meant it that way).

**Comment by filiphr** [comment]
@thunderhook, sorry about my confusing comment about the `@BeanMapping`. I was reviewing both PRs and I confused myself. What you are doing is correct. Your PR is as expected. I'll go ahead and merge it

**Comment by thunderhook** [comment]
You do not have to apologize. Thank you for your time reviewing and merging! 🫡
